cylc/flow/etc/job.sh

```diff
@@ -182,26 +233,41 @@ cylc__job__run_inst_func() {
 # Returns:
 #   exit 1
 cylc__job_finish_err() {
+    CYLC_TASK_USER_SCRIPT_EXITCODE="${CYLC_TASK_USER_SCRIPT_EXITCODE:-$?}"
     typeset signal="$1"
     typeset run_err_script="$2"
     shift 2
-    typeset signal_name=
     # (Ignore shellcheck "globbing and word splitting" warning here).
     # shellcheck disable=SC2086
     trap '' ${CYLC_VACATION_SIGNALS:-} ${CYLC_FAIL_SIGNALS}
     if [[ -n "${CYLC_TASK_MESSAGE_STARTED_PID:-}" ]]; then
         wait "${CYLC_TASK_MESSAGE_STARTED_PID}" 2>'/dev/null' || true
     fi
-    cylc message -- "${CYLC_SUITE_NAME}" "${CYLC_TASK_JOB}" "$@" || true
+    # Propagate real signals to entire process group, if we are a group leader,
+    # otherwise just to the backgrounded user script.
+    if [[ -n "${CYLC_TASK_USER_SCRIPT_PID:-}" ]] &&
+       [[ ":DEBUG:ERR:EXIT:RETURN:" != *":${signal}:"* ]]; then
+        kill -s "${signal}" -- "-$$" 2>'/dev/null' ||
+        kill -s "${signal}" -- "${CYLC_TASK_USER_SCRIPT_PID}" 2>'/dev/null' || true
+    fi
+    cylc message -- "${CYLC_SUITE_NAME}" "${CYLC_TASK_JOB}" "$@" &
```

Comment:
Time is of essence after receiving a signal, so we I think it's sensible to overlap network communication with `err-script`.